rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,308 @@
+  | 'rangeOverflow'
+  | 'rangeUnderflow'
+  | 'stepMismatch'
+  | 'tooLong'
+  | 'tooShort'
+  | 'typeMismatch'
+  | 'valid'
+  | 'valueMissing';
+```
+
+This means you can even show something whe the field is valid:
+
+```jsx
+<Form.Message match="valid">✅ Valid!</Form.Message>
+```
+
+#### Custom validation
+
+On top of all the built-in client-side validation matches described above you can also provide your own custom validation whilst still making use of the platform's validation abilities. It uses the `customError` type present in the constraint validity API.
+
+You can pass your own validation function into the `match` prop on `Form.Message`.
+
+```ts
+// The `match` prop also accepts a `CustomMatcher` (see below):
+type MatchProp = ValidityMatcher | CustomMatcher;
+type SyncCustomMatcher = (value: string, formData: FormData) => boolean;
+type AsyncCustomMatcher = (value: string, formData: FormData) => Promise<boolean>;
+```
+
+> `match` will be called with the current value of the control as first argument and the entire `FormData` as second argument.
+
+Here's a contrived example:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.Control />
+  <Form.Message match={(value, formData) => value !== 'John'}>Only John is allowed.</Form.Message>
+</Form.Field>
+```
+
+> `match` can also be an `async` function (or return a promise) to perform async validation (see type below):
+
+#### Messages outside of a `Form.Field`
+
+When nested inside a `Form.Field` part, `Form.Message` will automatically be wired to the control's validity which exists in the field.
+Alternatively you can render a message outside of a `Form.Field` part and pass it a `name` prop explicitly.
+
+#### Accessing the validity state for even more control
+
+Sometimes, you may need to access the raw validity state of a field in order to display your own icons, or interface with your own component library by passing certain props to it. You can do this by using the `Form.ValidityState` part:
+
+```jsx
+<Form.Field name="name">
+  <Form.Label>Full name</Form.Label>
+  <Form.ValidityState>
+    {(validity) => (
+      <Form.Control asChild>
+        <TextField variant="primary" state={getTextFieldState(validity)} />
+      </Form.Control>
+    )}
+  </Form.ValidityState>
+</Form.Field>
+```
+
+When nested inside a `Form.Field` part, `Form.ValidityState` will automatically be wired to the control's validity which exists in the field.
+Alternatively you can render it outside of a `Form.Field` part and pass it a `name` prop explicitly.
+
+#### Server-side validation
+
+The component also supports server-side validation using the same `Form.Message` component.
+You can re-use the same messages you defined for client-side errors by passing a `forceMatch` prop which will for the message to show regardless of the client-side matching logic. If the message doesn't exist on the client-side, you can render a `Form.Message` without a `match` too.
```

Comment:
Maybe it's just me, but I feel like the validation representation is a bit cluttered.

There are three main props involved in validation: `math`, `forceMath`, and `serverInvalid`, and I think it can be difficult to understand the relationship between them when first using this component.

- As described in the documentation, the `forceMath` prop can display messages regardless of client-side matching. While this is a useful feature, it can also be thought of as a sort of hack prop in that it "forces" something to be shown.
- If `serverInvalid` is false (not undefined) and `forceMatch` on `Form.Message` is true (if server-side or client-side validation results in an error), it's confusing whether the field evaluates as valid or invalid.

Why not limit the places where server-side validation can be handled to either `Form.Field` or `Form.Message`?

```jsx
const serverErrors = { missingValue: true, customError: true }

<Form.Field name="name" errors={serverErrors}>
  <Form.Label>Full name</Form.Label>
  <Form.Control />
    <Form.Message match="missingValue">
      Please enter your name.
    </Form.Message>
    <Form.Message match="customError">
      My Custom Error Message.
    </Form.Message>
</Form.Field>
```

Reply:
Yeah, @jjenzz has expressed similar concerns (allbeit maybe for different reasons) and we have chatted about it quite a bit. It might be nice to align closer to usual Radix components using uncontrolled/controlled props but it's not immediately clear how we may achieve that.

In the meantime, I think there's value in putting a release candidate in people's hands so you get to play with it and we can get more pointed feedback earlier.